File: hackerrank/problem solving/algorithms/2 - implementation/25 - save the prisoner/save_the_prisoner.go
// https://www.hackerrank.com/challenges/save-the-prisoner/problem?isFullScreen=true

package main

import "fmt"

func main() {
    var n int
    _, _ = fmt.Scan(&n)
    for range n {
        var prisoners, sweets, startChair int
        _, _ = fmt.Scan(&prisoners, &sweets, &startChair)
        fmt.Println(saveThePrisoner(prisoners, sweets, startChair))
    }
}

// T: O(1)
// S: O(1) extra space
func saveThePrisoner(prisoners, sweets, startChair int) int {
    return ((startChair - 1 + sweets - 1) % prisoners) + 1
}

File: hackerrank/problem solving/algorithms/2 - implementation/25 - save the prisoner/save_the_prisoner.rs
// https://www.hackerrank.com/challenges/save-the-prisoner/problem?isFullScreen=true

fn main() {
    let n: usize = text_io::read!();
    for _ in 0..n {
        let (prisoners, sweets, start_chair): (i32, i32, i32);
        text_io::scan!("{} {} {}", prisoners, sweets, start_chair);
        println!("{}", save_the_prisoner(prisoners, sweets, start_chair));
    }
}

// T: O(1)
// S: O(1) extra space
fn save_the_prisoner(prisoners: i32, sweets: i32, start_chair: i32) -> i32 {
    ((start_chair - 1 + sweets - 1) % prisoners) + 1
}
